### PR TITLE
JDK-8255439: System Tray icons get corrupted when windows scaling changes

### PR DESCRIPTION
**Pull request by honkar-jdk**
In Windows, when desktop scaling is changed the tray icons was distorted/blurred a bit each time scaling changes.

With the proposed fix, the tray icon scales according to on-the-fly DPI scale settings. A test case has been added which adds a MRI icon to system tray, to observe the icon scaling when DPI is changed. Since the scale cannot be programmatically changed (for dynamic on-the-fly scale changes), I have used a manual test case to test this scenario.

When DPI changes usually two messages are sent by windows -

- [WM_DPICHANGED](https://docs.microsoft.com/en-us/windows/win32/hidpi/wm-dpichanged)
- [WMPOSCHANGING](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-windowposchanging)

I'm triggering an update on tray icons on receiving WMPOSCHANGING msg through the Tray icon's Window Procedure. Triggering an update on WM_DPICHANGED was still causing the icons to be distorted, hence WMPOSCHANGING is being used as the message to trigger the update.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8255439](https://bugs.openjdk.java.net/browse/JDK-8255439): System Tray icons get corrupted when windows scaling changes

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - Author)
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Contributors
 * Alexey Ivanov `<aivanov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8441/head:pull/8441` \
`$ git checkout pull/8441`

Update a local copy of the PR: \
`$ git checkout pull/8441` \
`$ git pull https://git.openjdk.java.net/jdk pull/8441/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8441`

View PR using the GUI difftool: \
`$ git pr show -t 8441`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8441.diff">https://git.openjdk.java.net/jdk/pull/8441.diff</a>

</details>

### CODE REVIEW

**Comment by mrserb** [comment]
Did you have a chance to check why the WM_DPICHANGED does not work, any specific cases? It is used by checking the scale change for windows in AWT, and no issues were reported for that as far as I know.

**Comment by honkar-jdk** [comment]
@mrserb Wanted to follow up on this PR and previous feedback.

**Comment by honkar-jdk** [comment]
@mrserb  I wanted to add few more details about various window messages received during DPI changes. The order in which messages are received- 

1. WM_DPICHANGED - if update is triggered at this point, the tray icons get distorted
2. WM_POSCHANGING - if update is triggered at this point, tray icons are rendered without any distortion.
3. WM_POSCHANGED - message received when tray icon is removed or exited.

Kevin (@kevinrushforth) mentioned the usage of `WM_WINDOWPOSCHANGING` in JavaFX. Adding him to this PR conversation for suggestions.

**Review by kevinrushforth — COMMENTED** [review]
The fix looks good to me, but someone more familiar with the code will need to review it. I can confirm that the new test program fails without the fix and passes with the fix. One problem with the test is that it doesn't exit for me, regardless of whether the test passes or fails (or times out).

To partially answer Sergey's question, the problem with just using `WM_DPICHANGED` to trigger the update is that the scale of the default screen transform used by `WTrayIconPeer::updateNativeImage` hasn't yet been updated to reflect that new DPI scale. I don't know _why_ that is the case here, but it is. I do know that in JavaFX we stopped relying on `WM_DPICHANGED` for anything related to HiDPI several years ago when fixing some multi-monitor HiDPI issues in [JDK-8146920](https://bugs.openjdk.java.net/browse/JDK-8146920).

To show that triggering the update on `WM_DPICHANGED` is insufficient, I applied the patch from this PR, added some debugging output, and modified the fix to also update the image on a `WM_DPICHANGED` event. Here is the output when I start the program with a screen scale of 1.25 and then change it to 1.5:

```
WM_WINDOWPOSCHANGING
AwtTrayIcon::UpdateTrayIconHandler
SystemTray::updateTrayIcons
WTrayIconPeer::updateNativeImage : scale = (1.25, 1.25)
WM_DPICHANGED
AwtTrayIcon::UpdateTrayIconHandler
SystemTray::updateTrayIcons
WM_WINDOWPOSCHANGING
AwtTrayIcon::UpdateTrayIconHandler
SystemTray::updateTrayIcons
WTrayIconPeer::updateNativeImage : scale = (1.25, 1.25)  <----- This is from the WM_DPICHANGED event
WTrayIconPeer::updateNativeImage : scale = (1.25, 1.25)
WM_WINDOWPOSCHANGING
AwtTrayIcon::UpdateTrayIconHandler
SystemTray::updateTrayIcons
WTrayIconPeer::updateNativeImage : scale = (1.5, 1.5)    <----- Transform has now been updated
WM_WINDOWPOSCHANGING
AwtTrayIcon::UpdateTrayIconHandler
SystemTray::updateTrayIcons
WTrayIconPeer::updateNativeImage : scale = (1.5, 1.5)
WM_WINDOWPOSCHANGING
AwtTrayIcon::UpdateTrayIconHandler
SystemTray::updateTrayIcons
WTrayIconPeer::updateNativeImage : scale = (1.5, 1.5)
```

So I think the fix using `WM_WINDOWPOSCHANGING` to trigger the update is good. I left a couple minor comments inline.

**Comment by honkar-jdk** [comment]
@kevinrushforth Thank you for taking time to look into the fix and providing informative feedback. 
I'll check the test case exit issues and add the suggested changes.

**Comment by honkar-jdk** [comment]
Following are the approaches tried for Tray icon issue fix –

**1.	Using WM_DPICHANGED** – issue with this approach is the delay in scale values update as @kevinrushforth pointed out earlier.

**2.	Using DisplayChangedListener** -   To deal with delay in scale updates, @aivanov-jdk suggested to try DisplayChangedListener approach. This worked well, but tray icon distortion was still observed when the scales were changed in quick succession and especially when scaled down to 125%. 

Details of this approach -
•	Implemented a DisplayChangedListener on WTrayIconPeer 
•	Added displayChangedListener to local graphics environment to listen for DPI changes in WTrayIconPeer 
•	Override the displayChanged() so that WTrayIconPeer’s updateImage() is called whenever the DPI changes for the 
        default/main display.

**3.	Using WM_WINDOWPOSCHANGING** – With present testing, there is no distortion observed when an update to tray icon is triggered on this message.

**Comment by kevinrushforth** [comment]
Thanks for the additional comments. As far as I'm concerned, the only remaining question is the fact that the manual test doesn't exit for me when run from the command line.

**Comment by honkar-jdk** [comment]
> Thanks for the additional comments. As far as I'm concerned, the only remaining question is the fact that the manual test doesn't exit for me when run from the command line.

@kevinrushforth Thank you for the update Kevin. I'm currently looking into the test case exit issues when run from command line.

**Comment by aivanov-jdk** [comment]
Yes, with `WM_DPICHANGED` we faced the problem that Kevin describes: the update method gets called before the transform is updated. Thus it doesn't work.

Updating the transform, in its turn, calls `DisplayChangedListener`. It worked. Yet Harshitha's testing showed it wasn't reliable.

That's why she went for the third approach, which proved to work reliably. However, I haven't tested the latest version myself.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (aivanov-jdk) [inline test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:100]: You shouldn't call `System.exit` from a jtreg test.

The icon should probably be removed after pressing Pass or Fail button, this is likely what prevents the test from exiting.

The popup menu isn't important for this test case.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (aivanov-jdk) [inline test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:126]: Usually, files end with a new line.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (prrace) [inline test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:99]: Even if it is manual etc, we don't call System.exit() in jtreg tests .. we just shut down all the windows and dispose them so there is nothing to keep the main thread from exiting.
If this is called, I don't know why Kevin didn't see the app exit.
I suspect it cannot have been and he exited via some other path which left windows in existence which meant the toolkit thread didn't exit.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (prrace) [inline test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:72]: Hmm. When you get here, you already displayed the instructions window, and in the jtreg harness, the user will not see this message. They'll just not see the  tray icon and assume it failed.
I think you should be making use of the "forcePass()" code in PassFailJFrame

File: src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp
Comment (prrace) [inline src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:238]: Most of these JNI methods can leave an exception pending
https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#findclass
https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#getstaticmethodid
https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#callstatictypemethod-routines-callstatictypemethoda-routines-callstatictypemethodv-routines

I think you should be calling ExceptionClear() because subsequent methods should not be called with an exception pending. And this is a windows thread attached to the VM and it never "returns to Java" but will be called back to handle later events.

**Comment by honkar-jdk** [comment]
@prrace @aivanov-jdk Thank you for all the feedback and suggestions. I'll make sure to address all of them in the next update.

File: src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp
Comment (honkar-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:262]: @prrace So far, based on testing the message is received only for tray icon window. To be doubly sure, I'll test the scenarios when we receive this message on Tray Icon's Window procedure.

Comment (kevinrushforth) [reply src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:262]: > So far, based on testing the message is received only for tray icon window.

That matches what I see as well.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (kevinrushforth) [reply test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:99]: The `actionListener` is only called if you right click on the tray icon and select the Exit option. I didn't do that, so it never got here. Maybe the "Exit" button in the icon should be turned into a no-op, since you want the tester to press Pass or Fail to exit?

I think Alexey may have pointed out the problem with the test not exiting: it doesn't dispose the tray icon. That needs to be confirmed.

**Comment by aivanov-jdk** [comment]
Unfortunately, the icon is distorted on Windows 11. When I tested on Windows 10, it updates correctly most of the time, but rarely the icon becomes distorted. When the icon is distorted, enabling taskbar auto-hide and disabling it repaints the icon correctly; this works both on Windows 10 and 11.

Tray Icon 125% -> 150%
I got the above image on Windows 10 after switching from 125% to 150%. The icon should display 24 at 150%, one can see the number with some paint artefacts.

I don't understand why it happens this way.

**Comment by honkar-jdk** [comment]
Changes related to the test case and the test case exit issue is updated.

**Comment by honkar-jdk** [comment]
> Unfortunately, the icon is distorted on Windows 11. When I tested on Windows 10, it updates correctly most of the time, but rarely the icon becomes distorted. When the icon is distorted, enabling taskbar auto-hide and disabling it repaints the icon correctly; this works both on Windows 10 and 11.
> 
> Tray Icon 125% -> 150% I got the above image on Windows 10 after switching from 125% to 150%. The icon should display 24 at 150%, one can see the number with some paint artefacts.
> 
> I don't understand why it happens this way.

@aivanov-jdk Thank you for testing it on Win 10 & 11. I'm unable to replicate this issue on windows 10. But as you mentioned earlier, I was able to see similar icon distortion when using the DisplayChangedListener approach. I'll check what might be causing the issue on Win 11 and if it is replicable on Win 10 for any particular scenario.

**Comment by prrace** [comment]
> > Unfortunately, the icon is distorted on Windows 11. When I tested on Windows 10, it updates correctly most of the time, but rarely the icon becomes distorted. When the icon is distorted, enabling taskbar auto-hide and disabling it repaints the icon correctly; this works both on Windows 10 and 11.
> > Tray Icon 125% -> 150% I got the above image on Windows 10 after switching from 125% to 150%. The icon should display 24 at 150%, one can see the number with some paint artefacts.
> > I don't understand why it happens this way.
> 
> @aivanov-jdk Thank you for testing it on Win 10 & 11. I'm unable to replicate this issue on windows 10. But as you mentioned earlier, I was able to see similar icon distortion when using the DisplayChangedListener approach. I'll check what might be causing the issue on Win 11 and if it is replicable on Win 10 for any particular scenario.

@aivanov-jdk were both windows 10 and windows 11 using the same pipeline ? (meaning GDI or D3D).
Otherwise on our side I'd expect all the same code path ..

**Comment by aivanov-jdk** [comment]
> @aivanov-jdk were both windows 10 and windows 11 using the same pipeline ? (meaning GDI or D3D). Otherwise on our side I'd expect all the same code path ..

Good question. The output from `WindowsFlags` is the same on both machines:
```
WindowsFlags (Java):
  d3dEnabled: true
  d3dSet: false
  oglEnabled: false
  oglVerbose: false
  gdiBlitEnabled: true
  offscreenSharingEnabled: false
```
Setting `-Dsun.java2d.d3d=True` produces
```
Direct3D pipeline enabled on screen 0
```
in both cases.

Disabling Direct3D with `-Dsun.java2d.d3d=false` changes nothing.

**Comment by prrace** [comment]
>>Disabling Direct3D with -Dsun.java2d.d3d=false changes nothing.

That rules out a pipeline issue so I've no idea.
Fortunately Windows 11 still has lower adoption than Windows XP usage :-) That's a real stat btw.

**Comment by honkar-jdk** [comment]
I think the icon distortion observed on Win10 and 11 might be due to race condition between update triggered by WM_WINDOWPOSCHANGING and the re-addition of tray icons when taskbar is reloaded (occurs when DPI changes)

Whenever DPI changes, the taskbar is reloaded and all the existing tray icons are re-added by `WmTaskbarCreated()` in `awt_TrayIcon.cpp`. Here is the link to the relevant doc: [Taskbar creation and reload](https://docs.microsoft.com/en-us/windows/win32/shell/taskbar#taskbar-creation-notification) 

**The reason for distortion:** 
When update to tray icons (creating new scaled icon and deleting the old one) is called on `WM_WINDOWPOSCHANGING`, it might create race condition - updating the tray icon and deleting the previous ones while the  WmTaskbarCreated is trying to re-add the icons at the same time.

To overcome the race condition, I have removed the call to **UpdateTrayIconHandler()**  on WM_WINDOWPOSCHANGING and added it when taskbar gets reloaded and before `WmTaskbarCreated` is called.

When tested on Win10, it works as well. 

@aivanov-jdk Can you please test on Win10 & 11 with the new fix and let me know if it works?

**Comment by aivanov-jdk** [comment]
Yeah, this version works correctly on both test systems that I have.

I think the solution could be simplified. First of all, there's no need to call `UpdateTrayIconHandler` if DPI hasn't changed. A handler for `WM_DPICHANGED` sets a flag; then `WmTaskbarCreated` uses the flag to call updateImage.

`WmTaskbarCreated` iterates over all the icons, you can use this loop to call `WTrayIconPeer.updateImage`. This way all the icons will be updated on the toolkit thread. I'm not sure whether the handler should still sent both `NIM_MODIFY` and `NIM_ADD`; the former should be enough for DPI change (called from `updateNativeImage`), and the latter should be enough for the case where the Taskbar is really recreated.

**Comment by honkar-jdk** [comment]
@aivanov-jdk Thank you for looking into this issue and contributing a better & optimized approach. 
Additionally adding you as a contributor for this PR.

**Comment by honkar-jdk** [comment]
Based on @aivanov-jdk suggestions, new code changes have been incorporated. With the new change - tray icon update is triggered only when DPI change event is received and in WmTaskbarCreated().

Changes have been tested on Win 10 & 11 and works as expected.

**Review by kevinrushforth — COMMENTED** [review]
The latest version of the fix looks good, and is simpler than the earlier one. I tested it on Windows 10 and it works well.

There are two files that can be reverted, `TrayIcon.java` and `WTrayIconPeer.java`, since the only remaining changes after your last update are unrelated changes in the copyright year and the import statements. The easiest way to do this is `git checkout master -- filename1 filename2 ...`, presuming that the HEAD of `master` matches what you have merged into your branch.

**Comment by kevinrushforth** [comment]
One more thing to look at is that there is a failure in the GitHub Action run for the Windows x64 debug build that looks related to your changes. Make sure that you do a CI build with your latest changes.

File: src/java.desktop/share/classes/java/awt/TrayIcon.java
Comment (honkar-jdk) [reply src/java.desktop/share/classes/java/awt/TrayIcon.java:2]: @kevinrushforth Thank you for your suggestions . Will revert the files TrayIcon.java and WTrayIconPeer.java as suggested.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (honkar-jdk) [reply test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:129]: > One more thing to look at is that there is a failure in the GitHub Action run for the Windows x64 debug build that looks related to your changes. Make sure that you do a CI build with your latest changes.

Yes, there might have been a mix-up with variable name when I pushed the changes. Currently working on it. I'll check the CI build status before pushing the next update. Thank you!

File: src/java.desktop/windows/classes/sun/awt/windows/WTrayIconPeer.java
Comment (aivanov-jdk) [reply src/java.desktop/windows/classes/sun/awt/windows/WTrayIconPeer.java:2]: I'm for keeping the expanded imports in WTrayIconPeer.java, even it's the only change left now.

I'd probably remove the unused import from `TrayIcon.java` too. However, the change in this case is too minor.

Comment (kevinrushforth) [reply src/java.desktop/windows/classes/sun/awt/windows/WTrayIconPeer.java:2]: My preference would beto revert it. Once all of the actual changes associated with this fix were reverted, fixing the imports (or any similar type cleanup) in a file that otherwise doesn't need to be touched at all becomes an unrelated change. Maybe Phil can weigh in on this point.

Comment (aivanov-jdk) [reply src/java.desktop/windows/classes/sun/awt/windows/WTrayIconPeer.java:2]: Your approach makes the changeset _cleaner_, I agree. Nothing's changed in these two files. Better to avoid touching them and creating confusion. Let's leave these clean-up changes until the files are modified.

File: src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp
Comment (prrace) [inline src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:239]: Reduce this to ExceptionClear()
(1) ExceptionDescribe() per the docs https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#exceptions
prints a message .. which we don't want .. it is meant for debugging not production and FWIW it clears the pending exception so any time you see Describe() followed by Clear() the code seems odd (well pointless)
2) safe_ExceptionOccurred is only useful if you are coming from a JNI method that uses C++ exception handling. Here the call originates from native doesn't it ? And we don't return to Java since it is a native event loop.

So JUST env->ExceptionClear() will do

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (prrace) [reply test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:77]: > This should be the first statement in main. If icons aren't supported, there's no need to create the instruction frame. The test will still wait for user's response but no tray icon is added.

It won't wait if you haven't created it .. it will just exit without showing anything and be counted as pass, that's probably OK.

File: src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp
Comment (honkar-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:239]: @prrace Thanks for the details on ExceptionClear and its usage. Makes sense. I'll have it fixed on my next commit.

**Comment by honkar-jdk** [comment]
Changes incorporated:

- Added env->ExceptionClear() 
- Fixed CI debug build issue due to variable mismatch
- Added changes suggested by @prrace , @kevinrushforth and @aivanov-jdk

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (honkar-jdk) [reply test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:77]: Added the **!SystemTray.isSupported()** check as first line in main(). In case the system tray is not supported, the test is counted as pass and does not proceed to display instruction frame.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (honkar-jdk) [reply test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java:76]: Resolved by adding the check as first statement in main()

File: src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp
Comment (aivanov-jdk) [inline src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:73]: I'd rather drop 'Dpi' from the name because the name of the Java method is `updateImage`.

If you prefer, the C++ class method could also be changed to `UpdateImage`, dropping 'DPI'.

**Review by aivanov-jdk — APPROVED** [review]
I clean-built it and tested it, it works correctly on both Windows 10 and 11.

File: src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp
Comment (prrace) [reply src/java.desktop/windows/native/libawt/windows/awt_TrayIcon.cpp:73]: yes, that's a good point on the naming

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (aivanov-jdk) [inline test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java]: I'm nitpicking but I guess the blank line between the fields used in the test and instructions was in its own place to separate the two different groups.

File: test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java
Comment (aivanov-jdk) [inline test/jdk/java/awt/TrayIcon/TrayIconScalingTest.java]: There's usually a blank line between methods to visually separate the end of the previous method from the start of the following one.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

**Review by aivanov-jdk — APPROVED** [review]

**Comment by honkar-jdk** [comment]
@prrace This PR might require your re-review and re-approval on the latest commit.

**Review by prrace — APPROVED** [review]

**Review by kevinrushforth — APPROVED** [review]

**Comment by prrace** [comment]
type something to see if it wakes up the bots

**Pull request closed without merging**
